File: backend/src/main/java/com/apurv/auth/service/AuthService.java
package com.apurv.auth.service;

import java.time.Instant;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.apurv.auth.dto.AuthResponse;
import com.apurv.auth.dto.LoginRequest;
import com.apurv.auth.dto.RegisterRequest;
import com.apurv.auth.dto.TokenRefreshRequest;
import com.apurv.auth.entity.RefreshToken;
import com.apurv.auth.entity.Role;
import com.apurv.auth.entity.User;
import com.apurv.auth.repository.RefreshTokenRepository;
import com.apurv.auth.repository.UserRepository;
import com.apurv.auth.security.JwtService;
import com.apurv.common.exception.DuplicateResourceException;
import com.apurv.common.exception.ResourceNotFoundException;
import com.apurv.common.exception.TokenRefreshException;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    @Value("${jwt.expiration}")
    private long jwtExpiration;

    @Value("${jwt.refresh-expiration}")
    private long refreshTokenExpiration;

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final RefreshTokenRepository refreshTokenRepository;
    private final AuthenticationManager authenticationManager;
    private final TokenBlacklistService tokenBlacklistService;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new DuplicateResourceException("Email already in use");
        }

        if (userRepository.existsByUsername(request.getUsername())) {
            throw new DuplicateResourceException("Username already in use");
        }

        User user = User.builder()
                .email(request.getEmail())
                .username(request.getUsername())
                .password(passwordEncoder.encode(request.getPassword())).role(Role.MEMBER)
                .build();

        userRepository.save(user);
        log.info("New user registered: {}", user.getEmail());
        return buildAuthResponse(user);
    }

    @Transactional
    public AuthResponse login(LoginRequest request) {
        authenticationManager
                .authenticate(new UsernamePasswordAuthenticationToken(request.getUsername(), request.getPassword()));

        User user = userRepository.findByUsername(request.getUsername())
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        log.info("User logged in: {}", user.getUsername());
        return buildAuthResponse(user);
    }

    @Transactional
    public void logout(String authHeader, User currentUser) {
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            String token = authHeader.substring(7);
            long expirationTime = jwtService.getExpirationMs(token);
            if (expirationTime > 0) {
                tokenBlacklistService.blacklistToken(token, expirationTime);
            }
            refreshTokenRepository.deleteByUser(currentUser);
            log.info("User logged out: {}", currentUser.getUsername());
        }
    }

    @Transactional
    public AuthResponse refreshToken(TokenRefreshRequest request) {
        String requestRefreshToken = request.getRefreshToken();

        RefreshToken token = refreshTokenRepository.findByToken(requestRefreshToken)
                .orElseThrow(() -> new TokenRefreshException(requestRefreshToken, "Refresh token is not in database!"));

        if (token.isExpired()) {
            refreshTokenRepository.delete(token);
            throw new TokenRefreshException(requestRefreshToken,
                    "Refresh token was expired. Please make a new login request");
        }

        User user = token.getUser();
        AuthResponse response = buildAuthResponse(user);

        log.info("Access token refreshed for user: {}", user.getUsername());

        return response;
    }

    private AuthResponse buildAuthResponse(User user) {
        String accessToken = jwtService.generateToken(user);
        String refreshToken = saveRefreshToken(user);

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .userId(user.getId())
                .email(user.getEmail())
                .username(user.getUsername())
                .expiresIn(jwtExpiration / 1000)
                .build();
    }

    private String saveRefreshToken(User user) {
        refreshTokenRepository.deleteByUser(user);
        refreshTokenRepository.flush();

        RefreshToken refreshToken = RefreshToken.builder()
                .user(user)
                .token(jwtService.generateRefreshToken(user))
                .expiresAt(Instant.now().plusSeconds(refreshTokenExpiration / 1000))
                .build();

        refreshTokenRepository.save(refreshToken);
        return refreshToken.getToken();
    }
}

File: frontend/src/auth/service/AuthService.ts
import axios from "axios";
import api from "../../config/axios";
import type ApiResponse from "../../common/types/ApiResponse";
import type LoginRequest from "../types/request/LoginRequest";
import type RegisterRequest from "../types/request/RegisterRequest";
import type AuthResponse from "../types/response/AuthResponse";

function extractError(error: unknown, fallback: string): string {
  if (axios.isAxiosError<ApiResponse<unknown>>(error)) {
    return error.response?.data?.message || error.message || fallback;
  }
  if (error instanceof Error) return error.message;
  return fallback;
}

export async function login(req: LoginRequest) {
  try {
    const response = await api.post<ApiResponse<AuthResponse>>(
      "/auth/login",
      req,
    );
    if (response.data.success && response.data.data) {
      return { success: true as const, data: response.data.data };
    }
    return {
      success: false as const,
      error: response.data.message || "Login failed.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Login failed."),
    };
  }
}

export async function register(req: RegisterRequest) {
  try {
    const response = await api.post<ApiResponse<AuthResponse>>(
      "/auth/register",
      req,
    );
    if (response.data.success && response.data.data) {
      return { success: true as const, data: response.data.data };
    }
    return {
      success: false as const,
      error: response.data.message || "Registration failed.",
    };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Registration failed."),
    };
  }
}

export async function logout(refreshToken: string) {
  try {
    await api.post("/auth/logout", { refreshToken });
    return { success: true as const };
  } catch (error) {
    return {
      success: false as const,
      error: extractError(error, "Logout failed."),
    };
  }
}

const AuthService = {
  login,
  register,
  logout,
};

export default AuthService;
